docs/internal/interop.md

```diff
@@ -0,0 +1,351 @@
+# Verona Interoperability / Sandbox
+
+This is an initial attemp at mapping all the requirements that the
+interoperability layer will have in Verona to use external code written in other
+languages (like C++, Rust, etc) in a safe manner (sanbdox).
+
+## Usage in Verona
+
+The aim is to have external code in Verona with a native interface that looks
+and feels like Verona code. We want to abstract away the entirety of the sandbox
+functionality, the foreign function dispatch and serialisations, the execution
+models, etc.
+
+The syntax is still largely undefined, but this is an example of using the C++
+`std::vector` class inside Verona as pseudo-Verona code:
+
+```c
+// include/import the c++ code as a Verona module
+type StdVec = "c++::vector";
+
+// Create a new region (we know it's foreign, so it's a sandbox)
+reg = new StdVec;
+
+// Use Verona syntax to initialise a C++ vector
+obj = new std::vector[int] in reg
+
+// Use the object as if it was native
+obj.push_back(42)
+
+// The result will be used by a native Verona function
+print(obj.at(0))
+```
+
+The only hint that `obj` is a foreign object is in the region type being a
+sandbox. Not all foreign functionality will be available to the Verona code,
+especially in the beginning, so programmers are expected to create thin layers
+of foreign code to wrap functionality that can't be used in Verona (like what
+"extern C" does in C++).
+
+The syntax for foreign language code in Verona hasn't been defined yet, so we
+don't discuss it here. The pseudo-code in here is only a hint to start the
+discussion on the implementation.
+
+The objective of this document is to describe:
+
+* how is this going to interoperate with Verona and the target language
+* how the remote procedure calls will be dispatched and the arguments serialised
+* how the runtime will separate the memory from the main process and sanitise
+  the execution (system calls, file descriptors, ioctl, etc)
+
+## Overall architecture
+
+Each external language needs explicit support in the Verona compiler to be
+parsed at compile time and generate the correct stubs for the interoperability
+layer to correctly implement the remote procedure call (RPC).
+
+Each execution model (child process, web assembly, etc) will need explicit
+support for deploying the functionality at run time, hopefully not individually
+bound to the source language (N-to-M).
+
+But the overall architecture will be the same.
+
+### Parser
+
+1. Upon finding a file include that has foreign origin, the appropriate foreign
+   language parser (clang for C++) will be asked to parse the file and generate
+   its own AST.
+2. A Verona module will be created, with all foreign builtin types.
```

Comment:
We may want to surface this as a generic module that is imported by the specific library.  For example, a `verona.interop.c` that includes things like `float` and so on.  We may also want to make this come from the sandbox instance and make the modules generic on the sandbox instance (for example, a wasm sandbox `long` and a LP64 process-based sandbox `long` are not the same type).